projects/core/src/product/event/product-event.builder.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { filter, map, pairwise } from 'rxjs/operators';
+import { EventService } from '../../event/event.service';
+import { Breadcrumb } from '../../model/product-search.model';
+import { createFrom } from '../../util/create-from';
+import { ProductSearchService } from '../facade';
+import { FacetChangedEvent } from './product.events';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ProductEventBuilder {
+  constructor(
+    protected eventService: EventService,
+    protected productSearchService: ProductSearchService
+  ) {
+    this.register();
+  }
+
+  protected register() {
+    this.eventService.register(
+      FacetChangedEvent,
+      this.buildFacetChangedEvent()
+    );
+  }
+
+  protected buildFacetChangedEvent(): Observable<FacetChangedEvent> {
+    return this.productSearchService.getResults().pipe(
+      pairwise(),
+      filter(([prev, curr]) => {
+        if (prev && Object.keys(prev).length !== 0) {
+          const isCategory =
+            curr.breadcrumbs[0]?.facetCode === 'allCategories' &&
```

Comment:
Could you explain to me this logic I'm not sure I understand all the parts.

Reply:
When we select/unselect a facet, we must in the product search page. After toggle a facet, there is another product search request. So, we need 2 search results (`prev` and `curr`). 
To trigger the product search, either we do the `search,` or we go to the `category` page (route is `category` or `brands`). 
If it is `search`, we have to make sure `prev.freeTextSearch !== '' && prev.freeTextSearch === curr.freeTextSearch`.
If it is 'category`, we have to make sure the prev and curr's 'brand'/`category` is the same. For example, if in `Canon` brand, we have to make sure the new search result is also in `canon` brand.
```
curr.breadcrumbs[0]?.facetCode === 'allCategories' &&
            prev.breadcrumbs[0]?.facetCode === curr.breadcrumbs[0].facetCode &&
            prev.breadcrumbs[0].facetValueCode ===
              curr.breadcrumbs[0].facetValueCode;
```
the `prev.breadcrumbs[0]?.facetCode` and the `curr.breadcrumbs[0].facetCode` should be 'allCategories'. That means they are `category` (`category` has root `allCategories`).
the `prev.breadcrumbs[0].facetValueCode === curr.breadcrumbs[0].facetValueCode` should be the same. That means they are in the same brand or category. 
We toggle facet, the new request must be either the same search or in the same brand/category.